---
title: "imputeR Documentation"
author: "Jinliang Yang"
date: "October 8, 2015"
output:
  pdf_document:
    number_sections: yes
    toc: yes
  html_document:
    toc: yes
  word_document: default
---

\newpage

# Introduction

## Crossing Scheme and Experimental Design

![alt text](scheme.pdf)

In this experiment, we selfed and outcrossed a set of ~70 teosinte landraces to get a progeny array composed of 4,875 individuals. The ~70 founders and all the progeny were genotyped using GBS. We also re-sequenced 20/70 founder lines (the others will be re-sequenced soon). Because of the high error rate of the GBS data, especially problematic for calling heterozygous sites, we employed a phasing and imputation strategy to infer the expected genotypes by combining parentage and GBS information.

This file is to document the `R` package we developed to solve the problems step by step.

## Install and Usage

Install devtools first, and then use devtools to install imputeR from github.

```{r, eval=FALSE}
# install and load devtools
devtools::install_github("hadley/devtools")
library(devtools)
# install and load imputeR
install_github("yangjl/imputeR")
library(imputeR)
```

## How to find help.
Within "R" console, type `?impute_parent` or `help(impute_parent)` to find help information about the function.

```{r}
?impute_parent
```

# Infer parent's genotype from GBS data

If we got parent's WGS data, this step can be skipped.

We have observed parent and observed (selfed) kids.  We want to know $P(G|\theta)$, or the probability of parent's genotype given observed data $\theta$. And according to [Bayes' theorem](https://en.wikipedia.org/wiki/Bayes%27_theorem), 

$P(G|\theta) \propto P(\theta|G) \times P(G)$,   

where $P(G)$ is the probability of the genotype according to the Hardy-Weinberg equilibrium estimated from the population.
This consists of observed genotypes ($G'$) of both parent and kids. So:

$P(G|\theta)\propto \left( \prod\limits_{i=1}^{k}{P(G'_k|G)} \right) \times P(G'_{parent}|G) \times P(G)$,  

where $P(G'_{parent}|G)$ is the probability of parent's observed genotype given a true genotype $G$ by considering error rates, i.e. GBS homozygote error = 0.02 and heterozygote error = 0.8.  
And $P(G'_{k}|G)$ is the probability of the $k$ th kidâ€™s observed genotype given genotype $G$ by considering error rates and Mendelian segregation rate. The function `impute_parent` was implemented to compute parent's genotype probabilities.

## Toy example
In the below toy example, we simulated a full sib family with 12 kids for 3 loci. The observed parent's GBS genotype is `0 0 0` and kids are all `0 0 0`. In the resulting table, the first three columns are the probabilities of genotype `0, 1, 2`. The 4th column is the odd ratio of the highest divided by the 2nd highest probability. `gmax` parent's genotype with the highest probability. `gor` parent's genotype with the highest probability and `OR` bigger than the specified threshould.

```{r, eval=TRUE}
# parent's GBS data is a vector
obs_parent <- c(0, 0, 0)
# kids GBS data is a list of vectors
obs_kids <- list(c(2, 0, 0), c(0, 0, 0), c(2, 0, 0), c(0, 0, 0), c(2, 0, 0), c(1, 0, 0),
c(0, 0, 0), c(2, 0, 0), c(0, 0, 0), c(1, 1, 0), c(0, 0, 0), c(0, 0, 0))

# run impute_parent to get the probabilities of parent's genotype of all loci
geno <- impute_parent(obs_parent, obs_kids, hom.error=0.02, het.error=0.8, p=NULL)
# find the most likely genotype of parent
parentgeno(geno, oddratio=0.5, returnall=TRUE)
```

## Simulated Data

```{r, eval=FALSE}
set.seed(123456)
sim <- SimSelfer(size.array=10, het.error=0.8, hom.error=0.002, numloci=100, rec=1.2, imiss=0.3)
```

## Real Data

To load `hdf5` file, you could install Vince's [tasselr](https://github.com/vsbuffalo/tasselr) and [ProgenyArray](https://github.com/vsbuffalo/ProgenyArray) packages. And then, following the below instructions.


```{r, eval=FALSE}
# install tasselr and ProgenyArray, if you fail to install it, checking all the dependencies.
library(devtools)
install_github("vsbuffalo/tasselr")
install_github("vsbuffalo/ProgenyArray")

library(parallel)
options(mc.cores=NULL)
# you need to specify the location where the packages were installed. 
load_all("~/bin/tasselr")
load_all("~/bin/ProgenyArray")

# Note: at least 64G memory was needed to load the hdf5 file
loading_h5_recode(h5file="largedata/teo.h5", save.file="largedata/out.RData")    
```
Then, for each parent, run as above in the toy example. Note that unlike the toy example, you will need to supply a vector of allele frequencies at each locus estimated from the parents. If you do not supply this or leave `p=NULL`, a random allele frequency drawn from the neutral SFS will be used instead.  Since parents are coded as 0,1, or 2 for $N$ parents the allele frequency $p$ at a locus can be calculated as $\frac{\sum_{i=1}^Np_i}{2N}$.

----------


# Phasing Founder Genotypes  
$P(H|\theta) \propto P(\theta|H) \times P(H)$   
$P(H|\theta) \propto \left( \prod\limits_{i=1}^{k}{P(H'_k|H)} \right) \times P(H)$  
$P(H|\theta) \propto \left( \prod\limits_{i=1}^{k}\prod\limits_{l=1}^{n}{P(G'_{i,l}|H)} \right) \times P(H)$

- Where $\theta$ denotes observed data.
- $P(H)$ is the probability of the haplotype for a given window size of $n$.
- $P(G'_{i,l}|H)$ is the probability of kid $i$ at locus $l$ for a given haplotype $H$.
- The prior $P(H)$ is that all possible haplotypes of a given window size are equally likely.


----------

# Imputing and Phasing Kids  
$P(H_k|\theta) \propto P(\theta|H_k) \times P(H_k)$   
$P(H_k|\theta) \propto \left( \prod\limits_{i=k}{P(H'_k|H_k)} \right) \times P(H_k)$  
$P(H_k|\theta) \propto \left( \prod\limits_{i=k}\prod\limits_{l=1}^{n}{P(G'_{i,l}|H_k)} \right) \times P(H_k)$

- Where $\theta$ denotes observed data.
- $P(H)$ is the probability of the haplotype for a given window size of $n$.
- $P(G'_{i,l}|H)$ is the probability of kid $i$ at locus $l$ for a given haplotype $H$.
- The prior $P(H)$ is that all possible haplotypes of a given window size are equally likely.

```{r, eval=FALSE}
phase <- read.csv("../data/sim_phasing_res.csv")

hist(phase$er, breaks=30, main="Simulation (N=100)",col="#faebd7", xlab="Phasing Error Rate")
abline(v=mean(phase$er), col="red", lwd=2)
abline(v=median(phase$er), col="darkblue", lwd=2)
```

----------

# Phasing Dad of outcrossing progeny array  
$P(H_d|\theta) \propto P(\theta|H_d) \times P(H_d)$   
$P(H_d|\theta) \propto \left( \prod\limits_{i=1}^{k}{P(H'_k|H_d, H_m)} \right) \times P(H_m) \times P(H_d)$  
$P(H|\theta) \propto \left( \prod\limits_{i=1}^{k}\prod\limits_{l=1}^{n}{P(G'_{i,l}|H)} \right) \times P(H)$

- Where $\theta$ denotes observed data.
- $P(H_d)$ is the probability of the dad's haplotype for a given window size of $n$.
- $P(G'_{i,l}|H)$ is the probability of kid $i$ at locus $l$ for a given haplotype $H$.
- The prior $P(H)$ is that all possible haplotypes of a given window size are equally likely.


----------
